test/integration/apiserver/tracing/tracing_test.go

```diff
@@ -105,93 +84,461 @@ kind: TracingConfiguration
+	defer os.Remove(tracingConfigFile.Name())
+
+	if err := os.WriteFile(tracingConfigFile.Name(), []byte(fmt.Sprintf(`
+apiVersion: apiserver.config.k8s.io/v1alpha1
+kind: TracingConfiguration
+samplingRatePerMillion: 1000000
+endpoint: %s`, listener.Addr().String())), os.FileMode(0755)); err != nil {
+		t.Fatal(err)
+	}
 
-	traceFound := make(chan struct{})
-	defer close(traceFound)
 	srv := grpc.NewServer()
-	traceservice.RegisterTraceServiceServer(srv, &traceServer{
-		traceFound: traceFound,
-		filterFunc: containsNodeListSpan})
+	fakeServer := &traceServer{t: t}
+	fakeServer.resetExpectations([]*spanExpectation{})
+	traceservice.RegisterTraceServiceServer(srv, fakeServer)
 
 	go srv.Serve(listener)
 	defer srv.Stop()
 
 	// Start the API Server with our tracing configuration
 	testServer := kubeapiservertesting.StartTestServerOrDie(t,
 		kubeapiservertesting.NewDefaultTestServerOptions(),
-		apiserverArgs,
+		[]string{"--tracing-config-file=" + tracingConfigFile.Name()},
 		framework.SharedEtcd(),
 	)
 	defer testServer.TearDownFn()
-	clientConfig := testServer.ClientConfig
-
-	// Create a client that creates sampled traces.
-	tp := trace.TracerProvider(sdktrace.NewTracerProvider(sdktrace.WithSampler(sdktrace.AlwaysSample())))
-	clientConfig.Wrap(tracing.WrapperFor(tp))
-	clientSet, err := client.NewForConfig(clientConfig)
+	clientSet, err := client.NewForConfig(testServer.ClientConfig)
 	if err != nil {
 		t.Fatal(err)
 	}
 
-	// Make a request with the instrumented client
-	_, err = clientSet.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{})
-	if err != nil {
-		t.Fatal(err)
+	for _, tc := range []struct {
+		desc          string
+		apiCall       func(*client.Clientset) error
+		expectedTrace []*spanExpectation
+	}{
+		{
+			desc: "create node",
+			apiCall: func(c *client.Clientset) error {
+				_, err = clientSet.CoreV1().Nodes().Create(context.Background(),
+					&v1.Node{ObjectMeta: metav1.ObjectMeta{Name: "fake"}}, metav1.CreateOptions{})
+				return err
+			},
+			expectedTrace: []*spanExpectation{
+				{
+					name: "KubernetesAPI",
+					attributes: map[string]func(*commonv1.AnyValue) bool{
+						"http.user_agent": func(v *commonv1.AnyValue) bool {
+							return strings.HasPrefix(v.GetStringValue(), "tracing.test")
+						},
+						"http.target": func(v *commonv1.AnyValue) bool {
+							return v.GetStringValue() == "/api/v1/nodes"
+						},
+						"http.method": func(v *commonv1.AnyValue) bool {
+							return v.GetStringValue() == "POST"
+						},
+					},
+				},
+				{
+					name: "etcdserverpb.KV/Txn",
+					attributes: map[string]func(*commonv1.AnyValue) bool{
+						"rpc.system": func(v *commonv1.AnyValue) bool {
+							return v.GetStringValue() == "grpc"
+						},
+					},
+					events: []string{"message"},
+				},
+			},
+		},
+		{
+			desc: "get node",
+			apiCall: func(c *client.Clientset) error {
+				_, err = clientSet.CoreV1().Nodes().Get(context.Background(), "fake", metav1.GetOptions{})
+				return err
```

Comment:
I find this a little awkward that there is a dependency between these tests. I'm not sure if this is easy (or worth) to fix though

Reply:
mmm... I think you are right.  If I change the name, this Get() call fails.  I could ignore the error here, but then if the test fails I can't tell if the request failing was the reason or not.  I'd probably prefer to leave it as-is.

Reply:
That's good to know, yeah I was curious, because of how the tests are in an order that makes sense (create, get, ... delete). A different name would help, thanks!

Reply:
OK, Feel free to leave it as is, but maybe mention with a comment that this is order-dependant.

Reply:
added "This depends on the "create node" step having completed successfully" comments to steps that depend on creation.